File: src-tauri/src/database/chat_message.rs
use std::time::{SystemTime, UNIX_EPOCH};
use rusqlite::{Connection, params, Result, Row};
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Message {
    role: String,
    content: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ChatMessage {
    pub id: i64,
    pub create_time: i64,
    pub update_time: i64,
    pub messages: Vec<Message>,
    pub title: String,
}

fn get_chat_messages_from_query_result(row: &Row) -> ChatMessage {
    let messages: String = row.get(3).unwrap();
    let messages: Vec<Message> = serde_json::from_str(&messages).unwrap();
    ChatMessage {
        id: row.get(0).unwrap(),
        create_time: row.get(1).unwrap(),
        update_time: row.get(2).unwrap(),
        messages,
        title: row.get(4).unwrap(),
    }
}

pub fn init_chat_messages_table(conn: &Connection) -> Result<()> {
    conn.execute(
        "CREATE TABLE IF NOT EXISTS chat_message (
            id INTEGER PRIMARY KEY,
            create_time INTEGER NOT NULL,
            update_time INTEGER NOT NULL,
            messages TEXT,
            title TEXT
        )",
        [],
    )?;
    Ok(())
}

pub fn upgrade_chat_messages_table(conn: &Connection, _old_version: i64, _new_version: i64) -> Result<()> {
    let mut stmt = conn.prepare("SELECT sql FROM sqlite_master WHERE type = 'table' AND name = 'chat_message'")?;
    let mut rows = stmt.query([])?;
    if let Some(row) = rows.next()? {
        let sql: String = row.get(0)?;
        if !sql.contains("title") {
            conn.execute(
                "ALTER TABLE chat_message ADD COLUMN title TEXT",
                [],
            )?;
        }
    }
    Ok(())
}

pub fn create_chat_message(conn: &Connection, messages: Vec<Message>, title: String) -> Result<ChatMessage> {
    let mut stmt = conn.prepare("INSERT INTO chat_message (create_time, update_time, messages, title) VALUES (?1, ?2, ?3, ?4)")?;
    let create_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as i64;
    let messages_json = serde_json::to_string(&messages).unwrap();
    let insert_id = stmt.insert(params![create_time, create_time, messages_json, title])?;

    // 查询刚刚插入的数据
    let created_chat_message = get_chat_message_by_id(conn, insert_id).unwrap();

    Ok(created_chat_message)
}

pub fn update_chat_message(conn: &Connection, id: i64, messages: Vec<Message>, title: String) -> Result<ChatMessage> {
    let mut stmt = conn.prepare("UPDATE chat_message SET update_time = ?1, messages = ?2, title = ?3 WHERE id = ?4")?;
    let update_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as i64;
    let messages_json = serde_json::to_string(&messages).unwrap();
    stmt.execute(params![update_time, messages_json, title, id])?;
    let update_chat_message = get_chat_message_by_id(conn, id).unwrap();
    Ok(update_chat_message)
}

pub fn delete_chat_message(conn: &Connection, id: i64) -> Result<usize> {
    let mut stmt = conn.prepare("DELETE FROM chat_message WHERE id = ?1")?;
    let result = stmt.execute(params![id])?;
    Ok(result)
}

pub fn get_chat_message_by_id(conn: &Connection, id: i64) -> Result<ChatMessage> {
    let mut stmt = conn.prepare("SELECT id, create_time, update_time, messages, title FROM chat_message WHERE id = ?1")?;
    let mut rows = stmt.query(params![id])?;
    let row = rows.next()?.unwrap();
    Ok(get_chat_messages_from_query_result(&row))
}

pub fn get_chat_messages(conn: &Connection) -> Result<Vec<ChatMessage>> {
    let mut stmt = conn.prepare("SELECT id, create_time, update_time, messages, title FROM chat_message")?;
    let mut rows = stmt.query([])?;
    let mut chat_messages: Vec<ChatMessage> = Vec::new();
    while let Some(row) = rows.next()? {
        chat_messages.push(get_chat_messages_from_query_result(&row));
    }
    Ok(chat_messages)
}

File: src-electron/main/modules/tables/chat-message.ts
import Database from "better-sqlite3";
import { ChatSessionMessage, ChatMessage } from "@/types";

export default class ChatMessageTable {
  static initTable(db: Database.Database) {
    const createTableSql = `
      CREATE TABLE IF NOT EXISTS chat_message (
        id INTEGER PRIMARY KEY,
        create_time INTEGER NOT NULL,
        update_time INTEGER NOT NULL,
        messages TEXT,
        title TEXT,
        group_id INTEGER,
        archived INTEGER DEFAULT 0
      )
    `;
    db.exec(createTableSql);
  }

  static upgradeTable(db: Database.Database) {
    // 检查是否需要添加 group_id 字段
    const tableInfo = db
      .prepare("PRAGMA table_info(chat_message)")
      .all() as Array<{ name: string }>;
    const hasGroupId = tableInfo.some((col) => col.name === "group_id");
    const hasArchived = tableInfo.some((col) => col.name === "archived");

    if (!hasGroupId) {
      db.exec("ALTER TABLE chat_message ADD COLUMN group_id INTEGER");
    }

    if (!hasArchived) {
      db.exec("ALTER TABLE chat_message ADD COLUMN archived INTEGER DEFAULT 0");
    }

    // 添加索引
    try {
      db.exec(
        "CREATE INDEX IF NOT EXISTS idx_chat_message_group_id ON chat_message(group_id)",
      );
      db.exec(
        "CREATE INDEX IF NOT EXISTS idx_chat_message_archived ON chat_message(archived)",
      );
    } catch (e) {
      // 索引可能已存在
    }
  }

  static getListenEvents() {
    return {
      "create-chat-message": this.createChatMessage.bind(this),
      "update-chat-message": this.updateChatMessage.bind(this),
      "delete-chat-message": this.deleteChatMessage.bind(this),
      "get-chat-message-by-id": this.getChatMessageById.bind(this),
      "get-all-chat-messages": this.getAllChatMessages.bind(this),
    };
  }

  static parseChatMessage(chatMessage: {
    id: number;
    create_time: number;
    update_time: number;
    title: string;
    messages: string;
    group_id: number | null;
    archived: number;
  }): ChatMessage {
    return {
      id: Number(chatMessage.id),
      createTime: chatMessage.create_time,
      updateTime: chatMessage.update_time,
      title: chatMessage.title,
      messages: JSON.parse(chatMessage.messages),
      groupId: chatMessage.group_id,
      archived: Boolean(chatMessage.archived),
    };
  }

  static createChatMessage(
    db: Database.Database,
    messages: ChatSessionMessage[],
    title: string,
    groupId: number | null | undefined,
  ): ChatMessage {
    const stmt = db.prepare(
      "INSERT INTO chat_message (create_time, update_time, title, messages, group_id, archived) VALUES (?, ?, ?, ?, ?, 0)",
    );
    const now = Date.now();
    const res = stmt.run(
      now,
      now,
      title,
      JSON.stringify(messages),
      groupId ?? null,
    );
    const createdId = res.lastInsertRowid;
    return this.getChatMessageById(db, createdId);
  }

  static getChatMessageById(
    db: Database.Database,
    id: number | bigint,
  ): ChatMessage {
    const stmt = db.prepare("SELECT * FROM chat_message WHERE id = ?");
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return this.parseChatMessage(stmt.get(id) as any);
  }

  static updateChatMessage(
    db: Database.Database,
    chatMessage: Omit<ChatMessage, "updateTime">,
  ): ChatMessage {
    const { id, title, messages, groupId, archived } = chatMessage;
    const stmt = db.prepare(
      "UPDATE chat_message SET update_time = ?, title = ?, messages = ?, group_id = ?, archived = ? WHERE id = ?",
    );
    const now = Date.now();
    stmt.run(
      now,
      title,
      JSON.stringify(messages),
      groupId ?? null,
      archived ? 1 : 0,
      id,
    );
    return this.getChatMessageById(db, id);
  }

  static deleteChatMessage(db: Database.Database, id: number): number {
    const stmt = db.prepare("DELETE FROM chat_message WHERE id = ?");
    return stmt.run(id).changes;
  }

  static getAllChatMessages(db: Database.Database): ChatMessage[] {
    const stmt = db.prepare("SELECT * FROM chat_message");
    const chatMessages = stmt.all();
    return chatMessages.map((chatMessage) =>
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this.parseChatMessage(chatMessage as any),
    );
  }
}
